paddle/fluid/operators/partial_concat_op.cu

```diff
@@ -0,0 +1,220 @@
+    int64_t bs_index = id % out_batch_len;
+    int64_t var_id = bs_index / part_length;
+    int64_t part_index = bs_index % part_length;
+    int64_t in_id = start_index + part_index;
+    const T *tmp = in[var_id];
+    out[id] = tmp[bs_id * in_batch_len + in_id];
+    id += blockDim.x * gridDim.x;
+  }
+}
+
+template <class T>
+__global__ void ConcatPartialGradCUDAKernel(
+    T **in, const T *out, int64_t all_length, int64_t in_batch_len,
+    int64_t start_index, int64_t out_batch_len, int64_t part_length) {
+  int id = blockIdx.x * blockDim.x + threadIdx.x;
+  while (id < all_length) {
+    int64_t bs_id = id / out_batch_len;
+    int64_t bs_index = id % out_batch_len;
+    int64_t var_id = bs_index / part_length;
+    int64_t part_index = bs_index % part_length;
+    int64_t in_id = start_index + part_index;
+    T *tmp = in[var_id];
+    tmp[bs_id * in_batch_len + in_id] = out[id];
+    id += blockDim.x * gridDim.x;
+  }
+}
+
+template <typename T>
+class PartialConcatOpCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    auto in_vars = ctx.MultiInput<Tensor>("X");
+    Tensor *out = ctx.Output<Tensor>("Out");
+    PADDLE_ENFORCE_EQ(in_vars[0] != nullptr, true,
+                      platform::errors::InvalidArgument(
+                          "The input of partial concat should not be null."));
+
+    auto input_dim = in_vars[0]->dims();
+    PADDLE_ENFORCE_EQ(
+        input_dim.size(), 2,
+        platform::errors::InvalidArgument(
+            "Only supports 2-D array with "
+            "batch size in the 1st dimension and data in the 2nd."));
+    auto in_size = input_dim[1];
+    // may be negative
+    auto start_index = ctx.Attr<int>("start_index");
+    start_index = ComputeStartIndex(start_index, in_size);
+
+    auto partial_len = ctx.Attr<int>("length");
+    if (partial_len < 0) {
+      partial_len = in_size - start_index;
+    }
+
+    int in_num = in_vars.size();
+    int batch_size = input_dim[0];
+    int out_batch_len = partial_len * in_num;
+    int all_length = batch_size * out_batch_len;
+
+    constexpr size_t theory_sm_threads = 1024;
+    auto &dev_ctx = ctx.template device_context<platform::CUDADeviceContext>();
+    auto stream = dev_ctx.stream();
+    auto max_threads = dev_ctx.GetMaxPhysicalThreadCount();
+    auto sm_count = max_threads / theory_sm_threads;
+    size_t tile_size = 0;
+    dim3 grids;
+    dim3 blocks;
+    auto ComputeKernelParameter = [&](size_t length) {
+      if (length >= max_threads)
+        tile_size = 1024;
+      else if (length < max_threads && length > sm_count * 128)
+        tile_size = 512;
+      else if (length <= sm_count * 128)
+        tile_size = 256;
+      grids = dim3(CEIL_DIV(length, tile_size), 1, 1);
+      blocks = dim3(tile_size, 1, 1);
+    };
+
+    auto place = ctx.GetPlace();
+    T *out_data = out->mutable_data<T>(place);
+
+    std::vector<const T *> in_data;
```

Comment:
do not use dynamic vector, use size-fixed array: `const T** in_data[in_num];`

Reply:
Done.